src/config.js

```diff
@@ -0,0 +1,33 @@
+const mergeOptions = require('merge-options')
+const minimatch = require('minimatch')
+
+const getConfigForFunction = ({ config, func }) => {
+  if (!config) {
+    return {}
+  }
+
+  // It's safe to mutate the array because it's local to this function.
+  // eslint-disable-next-line fp/no-mutating-methods
+  const matches = Object.keys(config)
+    .filter((expression) => minimatch(func.name, expression))
+    .map((expression) => {
+      const wildcardCount = [...expression].filter((char) => char === '*').length
+
+      // The weight increases with the number of hardcoded (i.e. non-wildcard)
+      // characters — e.g. "netlify" has a higher weight than "net*". We do a
+      // subtraction of 1 if there is at least one wildcard character, so that
+      // "netlify" has a higher weight than "netlify*".
+      const weight = expression.length - wildcardCount - (wildcardCount ? 1 : 0)
+
+      return {
+        expression,
+        weight,
+      }
+    })
+    .sort((fA, fB) => fA.weight - fB.weight)
+    .map(({ expression }) => config[expression])
+
+  return mergeOptions.apply({ concatArrays: true }, matches)
```

Comment:
Will the `config` option include nested objects? If not, would a shallow merge work `Object.assign({}, ...matches)`?

Also, do we want to concatenate arrays instead of overriding? I think some users might expect simple overrides. Also, this removes the possibility for one globbing pattern to "undo" an array property from another globbing pattern.

Reply:
> Will the config option include nested objects?

I believe it will, yes. We've discussed the ability to define an `environment` object in the configuration objects, which would need to be deep-merged.

> Also, do we want to concatenate arrays instead of overriding? I think some users might expect simple overrides. Also, this removes the possibility for one globbing pattern to "undo" an array property from another globbing pattern.

In the case of external/ignored modules specifically, I think it's important that we concatenate. Imagine that the Next.js plugin specifies an array of external modules for all the plugin-generated functions. If you then add your own modules in the config, you would be unknowingly overriding what the plugin has defined and potentially breaking the build.

In general, I'd say that if someone defined a module as external, there's a specific reason behind it and we should be careful about overriding that behaviour. If we do think it's important that an expression is able to unset an external module defined by a less-specific glob, I would rather use the `!` notation (e.g. `externalModules: ["one", "!two"]`).

What do you think?

Reply:
Yes, this makes sense (on both points). :+1:

I am assuming `!` is already handled thanks to the globbing library, is that correct?

Reply:
> I am assuming ! is already handled thanks to the globbing library, is that correct?

Not yet. The globbing libary will operate on the function expressions alone, not on the values of the configuration properties. If we want to introduce support for the `!` syntax, we'll need to handle that ourselves.

Reply:
Should we create an issue to follow-up on this?

Reply:
Yes! https://github.com/netlify/zip-it-and-ship-it/issues/381

Reply:
> In general, I'd say that if someone defined a module as external, there's a specific reason behind it and we should be careful about overriding that behaviour. If we do think it's important that an expression is able to unset an external module defined by a less-specific glob, I would rather use the `!` notation (e.g. `externalModules: ["one", "!two"]`).

I think the negate approach is a bit problematic since users won't have any idea what external modules will be added by plugins, so they won't know which ones to negate.

How about a `force` syntax like we have with `redirects`?
This can work at a single property level (e.g. to avoid array concatenation) and at a glob pattern level (to avoid a merge).